internal/detector/kafkabroker/extract/attributes.go

```diff
@@ -0,0 +1,223 @@
+)
+
+const (
+	extProperties = ".properties"
+	flagOverride  = "--override"
+
+	// propertyLogDirs is the comma-separated paths to the Kafka logs.
+	propertyLogDirs = "log.dirs"
+	// propertyBrokerID is the broker ID for older versions of Kafka (Zookeeper)
+	propertyBrokerID = "broker.id"
+	// propertyNodeID is the equivalent of the propertyBrokerID in newer versions of Kafka (KRaft)
+	propertyNodeID = "node.id"
+	// propertyClusterID is the cluster ID.
+	propertyClusterID = "cluster.id"
+
+	logDirsSeparator    = ","
+	propertiesSeparator = '='
+	// fileNameMetaProperties is the generated properties file in the Kafka logs directory containing the cluster ID
+	// and is the source of truth for the broker ID.
+	fileNameMetaProperties = "meta.properties"
+
+	// brokerClassName is the main Java class name for the Kafka broker. This is what is normally run when starting a
+	// Kafka broker.
+	brokerClassName = "kafka.Kafka"
+)
+
+// brokerInfo holds identifying Kafka broker configuration fields extracted from command line arguments and properties
+// files.
+type brokerInfo struct {
+	serverProperties string
+	logDirs          string
+	brokerID         string
+}
+
+// ShouldParseProperties returns true if the properties file should be read to extract the log directories or broker ID.
+func (i *brokerInfo) ShouldParseProperties() bool {
+	return len(i.serverProperties) != 0 && !(i.HasLogDirs() && i.HasBrokerID())
+}
+
+func (i *brokerInfo) HasLogDirs() bool {
+	return len(i.logDirs) > 0
+}
+
+func (i *brokerInfo) HasBrokerID() bool {
+	return len(i.brokerID) > 0
+}
+
+type attributesExtractor struct {
+	logger *slog.Logger
+}
+
+func NewAttributesExtractor(logger *slog.Logger) detector.Extractor[map[string]string] {
+	return &attributesExtractor{
+		logger: logger,
+	}
+}
+
+// Extract Kafka broker attributes (broker.id and cluster.id) from a process. It parses the command line arguments to
+// identify Kafka brokers, reads configuration from the properties file if needed, and reads meta.properties files in
+// Kafka log directories for cluster information. Returns a map containing non-empty attribute values.
+func (e *attributesExtractor) Extract(ctx context.Context, process detector.Process) (map[string]string, error) {
+	args, err := process.CmdlineSliceWithContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	info, err := e.parseArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	// only parse the server properties if needed
+	if info.ShouldParseProperties() {
+		if err = e.parseServerPropertiesFile(ctx, process, info); err != nil {
+			e.logger.Debug("Failed to parse Kafka server properties", "pid", process.PID(), "err", err)
+		}
+	}
+
+	attributes := e.extractAttributesFromMetaProperties(ctx, process, info)
+	if _, ok := attributes[propertyBrokerID]; !ok && info.HasBrokerID() {
+		attributes[propertyBrokerID] = info.brokerID
+	}
+	for key, value := range attributes {
+		if value == "" {
```

Comment:
nit: Can we just do this before adding it above to avoid the extra iteration?

Reply:
Makes it a bit hard to do the early exit of the properties reading.